internal/sessioninterestmanager/sessioninterestmanager.go

```diff
@@ -40,26 +52,68 @@ func (sim *SessionInterestManager) RemoveSession(ses uint64) []cid.Cid {
 	sim.lk.Lock()
 	defer sim.lk.Unlock()
 
-	sim.wanted.RemoveSession(ses)
-	return sim.interested.RemoveSession(ses)
+	// The keys that no session is interested in
+	deletedKs := make([]cid.Cid, 0)
+
+	// For each known key
+	for c := range sim.wants {
```

Comment:
We'll need to profile this on the preload node and gateways. But we can merge it as-is.

Reply:
Note that this is the same thing that sessionWantlist was doing, but I agree it's a potential bottleneck. Hopefully all this code should run faster because we're no longer locking in both SessionInterestManager and sessionWantlist